### ISSUE TITLE
[docs] F.ctc_loss docs to warn clearly about invalid inf-causing inputs; zero_infinity to become enabled by default

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#31557]
**UPD** summary of all the long discussion for further discoverability:
1. F.ctc_loss will produce inf loss if presented with invalid unalignable examples
2. Such invalid examples may be generated by official usage code example if one is extremely unlucky or if one twists the dimension sizes a little bit
3. When presented with invalid examples, sum and mean reduction modes by default cause the whole batch loss to be **inf**

Proposals:
1. Have docs warn clearly about conditions on valid examples
2. Have docs warn clearly that the official usage example may produce invalid examples / fix the official code example
3. Enable zero_infinity = True by default or at least in reduction modes sum/mean

**BELOW IS THE ORIGINAL ISSUE DESCRIPTION**

```python
import torch
import torch.nn.functional as F

def test_ctc(C, B = 1, T = 2, full = False):
    log_probs = torch.randn(B, C, T).log_softmax(dim = 1)
    input_lengths = torch.full((B,), T, dtype=torch.long)

    target_lengths = torch.full((B,), T, dtype = torch.long)
    if full:
        targets = torch.full((B, T), C - 1, dtype = torch.long)
    else:
        targets = torch.randint(1, C, (B, T), dtype=torch.long)

    loss = F.ctc_loss(log_probs.permute(2, 0, 1), targets, input_lengths, target_lengths)
    print(float(loss))

test_ctc(C = 64, full = False)
# 4.894557952880859

test_ctc(C = 2, full = False)
# inf

test_ctc(C = 64, full = True)
# inf
```

Docs specify that targets can't be blank. If same consecutive labels are not supported, I think it should be explicitly mentioned in docs. And maybe docs should specify some workaround to encode consecutive same-valued targets (given that blank to separate them isn't allowed by docs).

Current docs: `Each element in the target sequence is a class index. And the target index cannot be blank (default=0)`

**Comment by mruberry** [comment]
Thanks for the suggestion! It'd be great if you wanted to submit a fix for this.

**Comment by vadimkantorov** [comment]
@mruberry One needs insight about CTC impl which I don’t have: can it be fixed or not. So in this case I can’t submit a fix.

**Comment by vadimkantorov** [comment]
This is potentially a bug in CTC. @mruberry could you please recategorize this?

**Comment by vadimkantorov** [comment]
@t-vi would you have any comments about proper encoding of repeated labels?

**Comment by t-vi** [comment]
The algorithm behind the function(s) is relatively vanilla the alignment but in logspace.
You never have a problem with target label repetitions, you just give it the target that you want. Only the input has the special encoding wrt repetitions and blanks.
My apologies for being blunt here, but based on the thread, I can recommend reading up on the ctc loss. Maybe the parameters become clearer after studying the original article or the distill pub exposition.

**Comment by vadimkantorov** [comment]
@t-vi For context: I've read the distill tutorial and other materials (including the original Graves paper) many times. I've been working on a speech recognition system for a few months and have had no problems with the CTC implementation in PyTorch (I used a special "repeat" character though).

As you know, I've been experimenting with an equivalent formulation of CTC in https://discuss.pytorch.org/t/manually-call-f-ctc-loss-backward-method/61634/6.

During the tests I've found that even the official PyTorch CTC docs example will print inf if the number of labels is reduced from 20 to 2 (i.e. we have 1-letter alphabet + blank char):
```python
import torch
import torch.nn.functional as F

torch.manual_seed(1)
log_probs = torch.randn(50, 16, 20).log_softmax(2).detach().requires_grad_()
targets = torch.randint(1, 2, (16, 30), dtype=torch.long)
input_lengths = torch.full((16,), 50, dtype=torch.long)
target_lengths = torch.randint(10,30,(16,), dtype=torch.long)
loss = F.ctc_loss(log_probs, targets, input_lengths, target_lengths)
print(loss)
# tensor(inf, grad_fn=<MeanBackward0>)
```

The only thing I did is to replace 20 by 2.

**Comment by t-vi** [comment]
This is not a bug but correct mathematical behaviour.

**Comment by vadimkantorov** [comment]
If this is correct mathematical behavior, shouldn't this aspect be documented? Increasing just number of target time steps in the official example also causes inf:
```python
import torch
import torch.nn.functional as F

torch.manual_seed(1)
log_probs = torch.randn(50, 16, 20).log_softmax(2).detach().requires_grad_()
targets = torch.randint(1, 20, (16, 50), dtype=torch.long)
input_lengths = torch.full((16,), 50, dtype=torch.long)
target_lengths = torch.randint(49,50,(16,), dtype=torch.long)
loss = F.ctc_loss(log_probs, targets, input_lengths, target_lengths)

print(loss)
# tensor(inf, grad_fn=<MeanBackward0>)
```

I will try to reimpl the forward pass in Python to try to understand the cause

**Comment by vadimkantorov** [comment]
@t-vi You are right. The situation seems to have nothing to do with repeated labels, but with the fact that input length 2 cannot possibly be aligned to a target sequence of two repeated labels. The behavior is mathematically correct. Btw I reimplemented 

Below are some proposals about changes to the docs:

How about adding to zero_infinity description sth like: "e.g. this can happen if aligning length-2 input to a sequence 'aa' ".

And maybe adding a comment in the code example:
`# this snippet does not check that alignment of inputs to targets is possible, so if you are super-unlucky you can get inf loss`

In the official code example if one is unfortunate to get 30 `a`'s as targets. 50-length input won't be enough to be alignable.

This is a bit confusing to receive inf on a good batch with one bad example - especially in default `sum` or `mean` reduction mode. A proposal: enable zero_infinity by default, at least with sum/mean reduction modes.

**Comment by t-vi** [comment]
I would have appreciated if you had clearly stated that you are wanting to feed inputs and targets such that the input sequence is mathematically impossible to align to the targets and this is why you get an infinite loss rather than being vague about it.
I'm sure there are bugs to be fixed and improvements to be made and it'd be good to improve the user experience, but I'm not certain that the suggestions in this issue would be on the top of my list. If someone else wants to submit a PR and find a reviewer to like it, then it is all well.

**Comment by vadimkantorov** [comment]
@t-vi If I had understood this root cause from the get-go, I would not have wasted everyone’s time! Please be ensured, I wrote this in good faith! The proof is I had posted a repro.

I twisted the official example very minorly and found it produces inf (and if one is extremely unlucky, even without any modifications it is theoretically possible to have it generate a bad example). Yes, the un-alignable examples are correlated with long examples which are correlated with repeated labels. Yes, my original cause analysis was false, but I did not invoke repeated labels on purpose to waste your time!

To be clear, I do not “want to feed inputs that are mathematically unalignable” per se, I only wish that docs are clear about the condition when this may happen (and this can happen by a minor twist of the official example!) and be fool-proof by default if this happens. 

At this point I hoped to discuss with you 1) need of doc fixes to clarify possibility of un-alignable examples from the official code example, 2) prospect of enabling zero_infinity by default, 3) idea of enabling zero_infinity by default when sum/mean reduction is used

Of course I do not expect anyone to fix them immediately (even if considered admissible), since it is clearly a not urgent or high-priority issue.

Sorry again for not finding the root cause myself from the get-go!

**Comment by vadimkantorov** [comment]
> I'm not certain that the suggestions in this issue would be on the top of my list.

Got it! I''ll rename the issue properly, so it's discoverable if someone hits it

**Comment by vadimkantorov** [comment]
@t-vi Could you please re-open the issue (for a discussion at some later point)?

I maybe will submit a docs PR at some point as well.
